Warmup

In [5]:
myList = [i for i in range(1, 1001) if i%2 == 0]
print(myList)

[2, 4, 6, 8, 10, 12, 14, 16, 18, 20, 22, 24, 26, 28, 30, 32, 34, 36, 38, 40, 42, 44, 46, 48, 50, 52, 54, 56, 58, 60, 62, 64, 66, 68, 70, 72, 74, 76, 78, 80, 82, 84, 86, 88, 90, 92, 94, 96, 98, 100, 102, 104, 106, 108, 110, 112, 114, 116, 118, 120, 122, 124, 126, 128, 130, 132, 134, 136, 138, 140, 142, 144, 146, 148, 150, 152, 154, 156, 158, 160, 162, 164, 166, 168, 170, 172, 174, 176, 178, 180, 182, 184, 186, 188, 190, 192, 194, 196, 198, 200, 202, 204, 206, 208, 210, 212, 214, 216, 218, 220, 222, 224, 226, 228, 230, 232, 234, 236, 238, 240, 242, 244, 246, 248, 250, 252, 254, 256, 258, 260, 262, 264, 266, 268, 270, 272, 274, 276, 278, 280, 282, 284, 286, 288, 290, 292, 294, 296, 298, 300, 302, 304, 306, 308, 310, 312, 314, 316, 318, 320, 322, 324, 326, 328, 330, 332, 334, 336, 338, 340, 342, 344, 346, 348, 350, 352, 354, 356, 358, 360, 362, 364, 366, 368, 370, 372, 374, 376, 378, 380, 382, 384, 386, 388, 390, 392, 394, 396, 398, 400, 402, 404, 406, 408, 410, 412, 414, 416, 418, 420, 42

In [6]:
myList2 = [int(i) for i in myList if '8' in str(i)]
print(myList2)

[8, 18, 28, 38, 48, 58, 68, 78, 80, 82, 84, 86, 88, 98, 108, 118, 128, 138, 148, 158, 168, 178, 180, 182, 184, 186, 188, 198, 208, 218, 228, 238, 248, 258, 268, 278, 280, 282, 284, 286, 288, 298, 308, 318, 328, 338, 348, 358, 368, 378, 380, 382, 384, 386, 388, 398, 408, 418, 428, 438, 448, 458, 468, 478, 480, 482, 484, 486, 488, 498, 508, 518, 528, 538, 548, 558, 568, 578, 580, 582, 584, 586, 588, 598, 608, 618, 628, 638, 648, 658, 668, 678, 680, 682, 684, 686, 688, 698, 708, 718, 728, 738, 748, 758, 768, 778, 780, 782, 784, 786, 788, 798, 800, 802, 804, 806, 808, 810, 812, 814, 816, 818, 820, 822, 824, 826, 828, 830, 832, 834, 836, 838, 840, 842, 844, 846, 848, 850, 852, 854, 856, 858, 860, 862, 864, 866, 868, 870, 872, 874, 876, 878, 880, 882, 884, 886, 888, 890, 892, 894, 896, 898, 908, 918, 928, 938, 948, 958, 968, 978, 980, 982, 984, 986, 988, 998]


In [9]:
from imblearn.over_sampling import RandomOverSampler

1. Look up SMOTE oversampling
https://imbalanced-learn.org/stable/references/generated/imblearn.over_sampling.SMOT E.html .
a. Describe what it is in your own words in markdown.
b. Use this technique with the diabetes dataset. Comment on the model
performance compared to other methods.

a.  SMOTE create new minority data points using original minority data points to better represent the minority class.  It doesn't repeat the same minoirty data points as naive oversampling.

In [11]:
import numpy as np
import pandas as pd

diabetes_df = pd.read_csv("diabetes.csv")
diabetes_df.head()

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   
2            8      183             64              0        0  23.3   
3            1       89             66             23       94  28.1   
4            0      137             40             35      168  43.1   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.627   50        1  
1                     0.351   31        0  
2                     0.672   32        1  
3                     0.167   21        0  
4                     2.288   33        1

In [12]:
from imblearn.over_sampling import SMOTE 
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

In [16]:
X = diabetes_df.drop('Outcome',axis = 1)
y = diabetes_df['Outcome']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 42, stratify = y)

#Standardize
sc = StandardScaler()
X_train_scaler = sc.fit_transform(X_train)
X_test_scaler = sc.fit_transform(X_test)

In [17]:
sm = SMOTE(random_state=42)

X_res, y_res = sm.fit_resample(X_train_scaler, y_train)

In [18]:
#train using the resampled data

model = LogisticRegression(random_state=42)
model.fit(X_res, y_res)


LogisticRegression(random_state=42)

In [19]:
#calculate the accuracy score
from sklearn.metrics import balanced_accuracy_score
y_pred = model.predict(X_test_scaler)
balanced_accuracy_score(y_test, y_pred)
# old resampling comment below
#increased accuracy from .6 to .74 by using oversampling

0.7268518518518519

Accuracy went to .73 - very close to other resampling techinque (naive)

Create a list of preprocessing steps you should try when working to build a model. Work
with your group to come up with the most comprehensive list you can.

* Correlation matrix
* Standard Scaler
* Log normal scaling
* One hot encoding
* Outlier detection
* Value counts
* Basic stats in .describe()
* Removing Nans or replacing them or ignoring them
* Clustering data
* New features based on current features
* Splitting dates/times
* Formatting column types
* Label encoding
* Integer encoding
* PCA - principal componant analysis
* Histograms for features